dialogue-client-verifier/src/test/java/com/palantir/verification/BinaryReturnTypeTest.java

```diff
@@ -0,0 +1,169 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.verification;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import com.google.common.collect.Iterables;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.palantir.conjure.java.api.config.service.ServiceConfiguration;
+import com.palantir.conjure.java.api.config.service.UserAgent;
+import com.palantir.conjure.java.api.config.service.UserAgents;
+import com.palantir.conjure.java.api.config.ssl.SslConfiguration;
+import com.palantir.conjure.java.client.config.ClientConfiguration;
+import com.palantir.conjure.java.client.config.ClientConfigurations;
+import com.palantir.conjure.java.dialogue.serde.DefaultConjureRuntime;
+import com.palantir.dialogue.Channel;
+import com.palantir.dialogue.example.SampleServiceAsync;
+import com.palantir.dialogue.hc4.ApacheHttpClientChannels;
+import io.undertow.Undertow;
+import io.undertow.server.HttpHandler;
+import io.undertow.server.HttpServerExchange;
+import io.undertow.util.Headers;
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Paths;
+import java.time.Duration;
+import java.util.Optional;
+import java.util.zip.GZIPOutputStream;
+import okhttp3.mockwebserver.MockResponse;
+import okhttp3.mockwebserver.MockWebServer;
+import okio.Buffer;
+import org.junit.Ignore;
+import org.junit.Test;
+
+public class BinaryReturnTypeTest {
+    private static final UserAgent USER_AGENT = UserAgent.of(UserAgent.Agent.of("foo", "1.0.0"));
+    private static final SslConfiguration SSL_CONFIG = SslConfiguration.of(
+            Paths.get("../dialogue-core/src/test/resources/trustStore.jks"),
+            Paths.get("../dialogue-core/src/test" + "/resources/keyStore.jks"),
+            "keystore");
+
+    @Test
+    public void undertow_server() {
+        Undertow undertow = Undertow.builder()
+                .addHttpListener(0, "localhost", new HttpHandler() {
+                    @Override
+                    public void handleRequest(HttpServerExchange exchange) throws IOException {
+                        exchange.getResponseHeaders().put(Headers.CONTENT_TYPE, "application/octet-stream");
+                        exchange.getResponseHeaders().put(Headers.CONTENT_ENCODING, "gzip");
```

Comment:
Does compression have anything to do with the problem, or does gzip expose the issue because otherwise the content fits into a buffer?

Reply:
The original stacktrace I saw made me think it was somehow the ContentDecodingChannel's fault so I tried to be as representative as possible:

```
Caused by: java.io.IOException: Attempted read on closed stream.
	at org.apache.http.conn.EofSensorInputStream.isReadAllowed(EofSensorInputStream.java:107)
	at org.apache.http.conn.EofSensorInputStream.read(EofSensorInputStream.java:116)
	at java.base/java.util.zip.CheckedInputStream.read(CheckedInputStream.java:60)
	at java.base/java.util.zip.GZIPInputStream.readUByte(GZIPInputStream.java:267)
	at java.base/java.util.zip.GZIPInputStream.readUShort(GZIPInputStream.java:259)
	at java.base/java.util.zip.GZIPInputStream.readHeader(GZIPInputStream.java:165)
	at java.base/java.util.zip.GZIPInputStream.<init>(GZIPInputStream.java:80)
	at com.palantir.dialogue.core.ContentDecodingChannel$DeferredGzipInputStream.getDelegate(ContentDecodingChannel.java:160)
	at com.palantir.dialogue.core.ContentDecodingChannel$DeferredGzipInputStream
```
In practise turning off all the content decoding & gzip stuff just gave me a different exception instead because the root cause was in the ConjureBodySerde class

Reply:
Got it, thanks for the description